src/Agent/NewRelic/Agent/Extensions/Providers/Wrapper/OpenAI/OpenAIChatWrapper.cs

```diff
@@ -0,0 +1,324 @@
+        {
+            segment.End();
+
+            if (responseTask.IsFaulted)
+            {
+                HandleError(segment, model, responseTask, agent);
+                return;
+            }
+
+            dynamic clientResult = GetTaskResult(responseTask);
+            if (clientResult == null)
+            {
+                agent.Logger.Log(Agent.Extensions.Logging.Level.Warn, $"Error processing response: Response payload is null");
+                return;
+            }
+
+            ProcessResponse(segment, model, chatMessages, clientResult, chatCompletionOptions, agent);
+        }
+    }
+
+    private string GetVendorName() => _isAzureOpenAI ? AzureOpenAIVendorName : OpenAIVendorName;
+
+    private string GetOrAddLibraryVersion(string assemblyFullName)
+    {
+        return _libraryVersions.GetOrAdd(assemblyFullName, VersionHelpers.GetLibraryVersion(assemblyFullName));
+    }
+
+    private static object GetTaskResult(object task)
+    {
+        if (((Task)task).IsFaulted)
+        {
+            return null;
+        }
+
+        var getResponse = _getResultFromGenericTask.GetOrAdd(task.GetType(), t => VisibilityBypasser.Instance.GeneratePropertyAccessor<object>(t, "Result"));
+        return getResponse(task);
+    }
+
+
+    private void ProcessResponse(ISegment segment, string model, dynamic chatMessages, dynamic clientResult, dynamic chatCompletionOptions, IAgent agent)
+    {
+        dynamic chatCompletionResponse = clientResult.Value;
+        string finishReason = chatCompletionResponse.FinishReason.ToString();
+
+        HandleSuccess(segment, model, chatMessages, clientResult, chatCompletionOptions, agent, chatCompletionResponse, finishReason);
+    }
+
+    private void HandleSuccess(ISegment segment, string requestModel, dynamic chatMessages, dynamic clientResult, dynamic chatCompletionOptions, IAgent agent, dynamic chatCompletionResponse, string finishReason)
+    {
+        _rolePropertyAccessor ??= VisibilityBypasser.Instance.GeneratePropertyAccessor<object>("OpenAI", "OpenAI.Chat.UserChatMessage", "Role");
+        _responseFieldGetter ??= VisibilityBypasser.Instance.GenerateFieldReadAccessor<object>("System.ClientModel", "System.ClientModel.ClientResult", "_response");
+
+        // typically there is only a single message in the outbound chat message list, but in a conversation, there can be multiple prompt and response message.
+        // The last message is the most recent prompt
+        // There can also be a refusal, which means there won't be a response message
+        dynamic lastChatMessage = chatMessages[chatMessages.Length - 1];
+
+        string refusal = chatCompletionResponse.Refusal;
+        string requestPrompt = refusal ?? lastChatMessage.Content[0].Text;
+
+        // roles need to be lowercase, but we're pulling the enum members by name so we need to lowercase them
+        string requestRole = _rolePropertyAccessor(lastChatMessage).ToString().ToLower();
```

Comment:
You can use the `EnumNameCache` for this, which also has a `GetNameToLower` method.

Reply:
I don't think `EnumNameCache<T>` will work on a dynamic since I don't know `T` at compile time.

Reply:
Ah you're right, my bad.